### PR TITLE
Always materialize the causal mask in Doge so sdpa stays causal

### LINKED ISSUE #48748: `DogeForCausalLM` attends to future tokens with the default `sdpa` backend — `SmallDoge/Doge-20M` scores ppl 5 instead of 16, and cached and uncached `generate()` disagree
### System Info

```
- `transformers` version: 5.18.0.dev0
- Platform: Linux-7.0.0-31-generic-x86_64-with-glibc2.39
- Python version: 3.10.20
- Huggingface_hub version: 1.31.0
- Safetensors version: 0.8.0
- Accelerate version: 1.13.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.12.1+cu130 (CUDA)
- Using distributed or parallel set-up in script?: no
- Using GPU in script?: no, the script below is CPU-only
- GPU type: NVIDIA GeForce RTX 3060
```

### Who can help?

@vasqu @Cyrilvallez

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Scoring `SmallDoge/Doge-20M` (the docs-page checkpoint), the perplexity changed with `attn_implementation`. Under the default `sdpa` backend the model is not causal: logits at position *i* depend on tokens after *i*. `eager` is fine.

`main` at `415e6d2f59`, CPU:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

name = "SmallDoge/Doge-20M"
tok = AutoTokenizer.from_pretrained(name)
text = tok("Paris is the capital and most populous city of France, with an estimated population of over two million residents.", return_tensors="pt")
prompt = tok("The capital of France is", return_tensors="pt")

for impl in ("eager", "sdpa"):
    model = AutoModelForCausalLM.from_pretrained(name, attn_implementation=impl, dtype=torch.float32).eval()
    with torch.no_grad():
        # teacher-forced loss
        loss = model(**text, labels=text["input_ids"]).loss
        # do earlier logits change when only the last token changes?
        other = {k: v.clone() for k, v in text.items()}
        other["input_ids"][0, -1] = 13
        shift = (model(**text).logits[0, :-1] - model(**other).logits[0, :-1]).abs().max()
        # greedy decoding with and without the cache
        cached = tok.decode(model.generate(**prompt, max_new_tokens=12, do_sample=False)[0], skip_special_tokens=True)
        uncached = tok.decode(model.generate(**prompt, max_new_tokens=12, do_sample=False, use_cache=False)[0], skip_special_tokens=True)
    print(f"{impl:5} loss {loss.item():.3f} (ppl {loss.exp().item():.1f}) | earlier logits move by {shift.item():.3f} | cached == uncached: {cached == uncached}")
```

```
eager loss 2.779 (ppl 16.1) | earlier logits move by 0.000 | cached == uncached: True
sdpa  loss 1.613 (ppl 5.0) | earlier logits move by 3.911 | cached == uncached: False
```

A tiny random `DogeConfig` shows the same. It happens whenever `create_causal_mask` takes the sdpa shortcut and returns `None` — unpadded input, an all-ones `attention_mask`, or any cached call — and disappears once a pad token forces the mask to be built (sdpa then matches eager).

Cause: `DogeAttention.prepare_dynamic_mask` folds the causal mask into its dense `dt_states` bias only when handed one (`modeling_doge.py:359-365`) and always passes that bias on as `attention_mask`. Given `None`, the bias reaches sdpa with no causal term, and `is_causal` is off because a mask was supplied.

Not new: the code has been this way since Doge was added (#35891) for any tokenizer mask or cached call; #42367 extended it to the bare call. Nobody has reported it — `has_attentions = False` hides it from the eager-vs-sdpa tests, and the slow `test_Doge_20M_hard` golden pins the sdpa text.

### Expected behavior

`sdpa` and `eager` should agree, as they do once a padding mask is present: a change to token *k* leaves the logits before *k* untouched, and cached and uncached `generate()` agree.

Who this hits: anyone evaluating or fine-tuning Doge with default settings — the loss is too good because the labels are visible — and anyone generating without the cache. The remote-code `modeling_doge.py` in the `Doge-20M` repo always builds the mask.

***I have a fix (one argument at the mask call, the pattern Falcon and the DSA masks use), a regression test and the regenerated golden ready locally — red on `main`, green with the fix. Happy to open the PR if wanted.***

### PR DESCRIPTION
**Pull request by blipbyte**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48821) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48821)

## What does this PR do?

Fixes #48748.

`DogeAttention.prepare_dynamic_mask` folds the causal mask into its dense `dt_states` bias only when the mask is not `None` (`modeling_doge.py:359-365`), and always passes that bias on as `attention_mask`. Under sdpa, `create_causal_mask` returns `None` whenever the mask would be plain causal — unpadded input, an all-ones mask, any cached call — so the bias reaches sdpa with no causal term and `is_causal` is off because a mask was given.

On `main`, `SmallDoge/Doge-20M`: loss 1.613 (ppl 5.0) under sdpa vs 2.779 (ppl 16.1) under eager; the last token moves earlier logits by 3.9; cached and uncached greedy `generate()` disagree. Fixed, all three match eager. Present since #35891 for any tokenizer mask or cached call; #42367 extended it to the bare call.

The fix is `allow_is_causal_skip=False` at the mask call — what its docstring prescribes for a mask "later concatenated with another mask", as in `modeling_falcon.py:769` and the DSA / Qwen4 mask entries (#48456) — in `DogeModel.forward` and in a `create_masks_for_generate` override on `DogeForCausalLM`, the builder `generate` uses ahead of a static cache (DSA gets it through the layer-type mapping; Doge has no `layer_types`). The seven other models that fold this mask into their own bias all handle `None`.

Tests. `has_attentions = False` keeps the eager-vs-sdpa suite off for Doge and the cache-parity tolerance hides the drift at tiny init, so `create_and_check_sdpa_decoder_is_causal` asserts the invariant under sdpa for the three shapes that leaked: no mask, all-ones mask, with cache. Red on `main`, green with the fix; no existing test reddens when the fix is reverted. The override changes no output (the forward re-materializes what `generate` skipped), so no test guards it alone.

`test_Doge_20M_hard`'s golden was generated under sdpa and pinned the bidirectional text; it is regenerated to eager's output on `main`, identical for fixed sdpa on CPU bf16 and RTX 3060 bf16. The old `("cuda", 8)` string already failed on that GPU.

```
pytest tests/models/doge/   CPU main 110 passed · branch 111 passed, 0 failed · GPU main 116 · branch 117, 0 failed
RUN_SLOW=1 test_Doge_20M_hard: passed (RTX 3060)
make fix-repo 17/17 · make check-repo 26/26 · make typing · make style
```

Not in this PR: flipping `has_attentions` — green after the fix, but those 24 eager-vs-sdpa tests also pass on `main` with the bug, so it is a cleanup, not a guard.

AI-assisted, per `CONTRIBUTING.md`. Coordination: #48748, which I opened; PR at @vasqu's request there. Differentiation: #48769 / #48770 add the same argument; this one also regenerates the slow golden pinned to the old sdpa output, covers the cached call, and forces the generate-side mask. I reviewed every changed line and ran the suites locally.

## Code Agent Policy

- [ ] (First-time contributors only): I confirm that this PR description and code is not written by an LLM or code agent

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #48748
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@vasqu (attention masks, #42367 / #48456) @Cyrilvallez

### CODE REVIEW

File: src/transformers/models/doge/modeling_doge.py
Comment (vasqu) [inline src/transformers/models/doge/modeling_doge.py:1]: I guess we also need to override the mask creation for generate no? So `create_masks_for_generate` under `ForCausalLM`

File: tests/models/doge/test_modeling_doge.py [hunk 1]
```diff
@@ -336,6 +336,23 @@ def test_doge_sequence_classification_model_for_multi_label(self):
     def test_save_load_fast_init_from_base(self):
         pass
 
+    def test_sdpa_decoder_is_causal(self):
```
Comment (vasqu) [inline tests/models/doge/test_modeling_doge.py:355]: can you check the fast test design in mamba2, i.e. the tester does most of the work

File: tests/models/doge/test_modeling_doge.py [hunk 2]
```diff
@@ -336,6 +336,23 @@ def test_doge_sequence_classification_model_for_multi_label(self):
...
+        config, inputs = self.model_tester.prepare_config_and_inputs_for_common()
+        input_ids = inputs["input_ids"]
+        other_input_ids = input_ids.clone()
+        other_input_ids[:, -1] = (input_ids[:, -1] + 1) % config.vocab_size
+        model = DogeForCausalLM._from_config(config, attn_implementation="sdpa").to(torch_device).eval()
```
Comment (vasqu) [inline tests/models/doge/test_modeling_doge.py]:
```suggestion
        model = DogeForCausalLM(config).to(torch_device).eval()
        model.set_attn_implementation("sdpa")
```

File: tests/models/doge/test_modeling_doge.py [hunk 3]
```diff
@@ -336,6 +336,23 @@ def test_doge_sequence_classification_model_for_multi_label(self):
...
+        cases = {
+            "no mask": {"use_cache": False},
+            "all-ones mask": {"attention_mask": torch.ones_like(input_ids), "use_cache": False},
+            "with cache": {"use_cache": True},
+        }
+        for name, kwargs in cases.items():
+            with self.subTest(name), torch.no_grad():
```
Comment (vasqu) [inline tests/models/doge/test_modeling_doge.py]: no need for subtests

File: tests/models/doge/test_modeling_doge.py [hunk 4]
```diff
@@ -373,8 +390,7 @@ def test_Doge_20M_hard(self):
         """
         EXPECTED_TEXT = Expectations(
             {
-                (None, None): "Here's everything I know about dogs. Dogs is the best animal in the world. It is a very popular and popular dog in the United States. It is a very popular",
-                ("cuda", 8): "Here's everything I know about dogs. Dogs is the best animal in the world. It is a very popular and popular breed for dogs. It is a very popular and popular",
+                (None, None): "Here's everything I know about dogs. Dogs is the best animal in the world, and they are the most common pets. Dogs are known for their unique personalities and behaviors,",
```
Comment (vasqu) [inline tests/models/doge/test_modeling_doge.py:396]: will need to adjust the cuda 8 case then but no worries let's wrap up the general stuff first

**Review by vasqu — COMMENTED** [review]
Some initial comments 🤗

File: src/transformers/models/doge/modeling_doge.py
Comment (blipbyte) [reply src/transformers/models/doge/modeling_doge.py:1]: Good call, added as a `@staticmethod` on `DogeForCausalLM` in the modular file, same shape as the FalconH1 / NemotronH ones: it picks the same mask function the forward does and passes `allow_is_causal_skip=False`.

For the record, what it changes: with `cache_implementation="static"` under sdpa, the generic `create_masks_for_generate` returned `None` at prefill (decode steps were already materialized by the `is_compileable and q_length == 1` rule in `create_causal_mask`), and the forward's forced mask caught it, so greedy output already matched eager on this branch for a single prompt and a left-padded batch. With the override, `generate` now hands the prefill a 4D mask instead of `None`, same as the DSA entries do through the layer-type mapping. Output is unchanged; the point is that both mask builders now agree.

File: tests/models/doge/test_modeling_doge.py (hunk 1 above)
Comment (blipbyte) [reply tests/models/doge/test_modeling_doge.py:355]: Done, moved into `DogeModelTester.create_and_check_sdpa_decoder_is_causal` with the two-line `test_` wrapper, like `test_mamba2_caching`.

File: tests/models/doge/test_modeling_doge.py (hunk 4 above)
Comment (blipbyte) [reply tests/models/doge/test_modeling_doge.py:396]: Sure. The only cc-8.x GPU I have is an RTX 3060 (8.6), and there the fixed output equals the CPU string, which is why the entry went away rather than getting a new value. Happy to add whatever the run-slow on the CI card produces, or to put the 3060 string back under `("cuda", 8)` if you would rather keep the key.

File: tests/models/doge/test_modeling_doge.py (hunk 3 above)
Comment (blipbyte) [reply tests/models/doge/test_modeling_doge.py]: Dropped; it is a plain loop over the three inputs now, the assertion message names which one failed.

File: tests/models/doge/test_modeling_doge.py (hunk 2 above)
Comment (blipbyte) [reply tests/models/doge/test_modeling_doge.py]: Taken as suggested (the lines moved into the tester, so I applied it by hand).

**Comment by blipbyte** [comment]
Addressed the comments in fdb33aa7c0: `create_masks_for_generate` override on `DogeForCausalLM`, test moved into the tester without subtests and using `set_attn_implementation`. The `("cuda", 8)` golden is left for the run-slow as discussed.

File: tests/models/doge/test_modeling_doge.py (hunk 4 above)
Comment (vasqu) [reply tests/models/doge/test_modeling_doge.py:396]: Yup no worries will adjust it tomorrow

**Review by vasqu — APPROVED** [review]

**Comment by vasqu** [comment]
run-slow: doge

**Comment by vasqu** [comment]
Sorry for the delays and congrats on the merge :hugs:
